tsl/src/compression/compression.c

```diff
@@ -2239,118 +2236,559 @@ compressed_insert_key_columns(Relation relation)
+								   attnum,
+								   BTEqualStrategyNumber,
+								   InvalidOid, /* No strategy subtype. */
+								   index_rel->rd_att->attrs[AttrNumberGetAttrOffset(attnum)]
+									   .attcollation,
+								   opr,
+								   value);
+		}
+
+		if (matches)
+		{
+			*result_index_rel = index_rel;
+			break;
+		}
+		else
+		{
+			index_close(index_rel, AccessShareLock);
+			pfree(scankeys);
+			scankeys = NULL;
+		}
+	}
+
+	return scankeys;
+}
+
+static void
+report_error(TM_Result result)
+{
+	switch (result)
 	{
+		case TM_Deleted:
+		{
+			if (IsolationUsesXactSnapshot())
+			{
+				/* For Repeatable Read isolation level report error */
+				ereport(ERROR,
+						(errcode(ERRCODE_T_R_SERIALIZATION_FAILURE),
+						 errmsg("could not serialize access due to concurrent update")));
+			}
+		}
+		break;
 		/*
-		 * If there are no unique constraints there is nothing to do here.
+		 * If another transaction is updating the compressed data,
+		 * we have to abort the transaction to keep consistency.
 		 */
-		return;
+		case TM_Updated:
+		{
+			elog(ERROR, "tuple concurrently updated");
+		}
+		break;
+		case TM_Invisible:
+		{
+			elog(ERROR, "attempted to lock invisible tuple");
+		}
+		break;
+		case TM_Ok:
+			break;
+		default:
+		{
+			elog(ERROR, "unexpected tuple operation result: %d", result);
+		}
+		break;
 	}
+}
 
-	if (!ts_guc_enable_dml_decompression)
-		ereport(ERROR,
-				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-				 errmsg("inserting into compressed chunk with unique constraints disabled"),
-				 errhint("Set timescaledb.enable_dml_decompression to TRUE.")));
+static inline TM_Result
+delete_compressed_tuple(RowDecompressor *decompressor, Snapshot snapshot,
+						HeapTuple compressed_tuple)
+{
+	TM_FailureData tmfd;
+	TM_Result result;
+	result = table_tuple_delete(decompressor->in_rel,
+								&compressed_tuple->t_self,
+								decompressor->mycid,
+								snapshot,
+								InvalidSnapshot,
+								true,
+								&tmfd,
+								false);
+	return result;
+}
 
-	Oid comp_relid = ts_chunk_get_relid(cis->compressed_chunk_id, false);
-	Relation in_rel = relation_open(comp_relid, RowExclusiveLock);
+struct decompress_batches_stats
+{
+	int64 batches_decompressed;
+	int64 tuples_decompressed;
+};
 
-	RowDecompressor decompressor = build_decompressor(in_rel, out_rel);
-	Bitmapset *key_columns = compressed_insert_key_columns(out_rel);
-	Bitmapset *null_columns = NULL;
+/*
+ * This method will:
+ *  1.Scan the index created with SEGMENT BY columns.
+ *  2.Fetch matching rows and decompress the row
+ *  3.insert decompressed rows to uncompressed chunk
+ *  4.delete this row from compressed chunk
+ *
+ *  Returns whether we decompressed anything.
+ */
+static struct decompress_batches_stats
+decompress_batches_using_index(Relation in_rel, Relation out_rel, Relation index_rel,
+							   Snapshot snapshot, ScanKeyData *index_scankeys,
+							   int num_index_scankeys, ScanKeyData *scankeys, int num_scankeys,
```

Comment:
Suggest `heap_scankeys`, `num_heap_scankeys` to avoid confusion with "all scankeys".